File: routes/studio.py
"""Studio integration — connection state, attention inbox, deep links.

All Studio I/O happens on the background refresher thread started in serve.py,
which writes an in-process snapshot. Request handlers (including /poll) only
ever READ that snapshot, so a slow or dead Studio can never stall the UI.
"""
import copy
import os
import re
import subprocess
import threading
import time
from pathlib import Path

from flask import Blueprint, jsonify, request

from routes.poll import _find_project_dir
from shared import state
from shared.studio_client import client

studio_bp = Blueprint("studio", __name__)

# Refresher cadence. Inbox is the badge's freshness budget; health is cheap but
# pointless to hammer; a disconnected Studio backs off so a dead host isn't
# probed six times a minute forever.
_INBOX_INTERVAL = 10.0
_HEALTH_INTERVAL = 30.0
_BACKOFF_INTERVAL = 60.0

# Sessions launched by `sto session start` are named studio_e<effort>_s<session>.
_SESSION_EFFORT_RE = re.compile(r"^studio_e(\d+)_s\d+$")

_snapshot = {
    "state": "unconfigured",
    "inbox": [],
    "inbox_count": 0,
    "checked_at": 0.0,
    "http_code": None,
}
_snapshot_lock = threading.Lock()
# Only one refresh may run at a time: the refresher thread and a concurrent
# /studio/test would otherwise publish their snapshots out of order.
_refresh_lock = threading.Lock()
# question_id -> time answered. An inbox fetch that began BEFORE an answer
# committed is still in flight; without this it republishes the answered
# question and resurrects the badge for a whole cycle.
_answered_recently = {}
_ANSWERED_TTL = 120.0


def _prune_answered():
    cutoff = time.time() - _ANSWERED_TTL
    for qid in [q for q, at in _answered_recently.items() if at < cutoff]:
        _answered_recently.pop(qid, None)


def _publish_inbox_locked(rows, now):
    """Store an inbox fetch into the snapshot. Caller must hold _snapshot_lock.

    Every path that publishes an inbox goes through here so the
    _answered_recently filter cannot be skipped: a fetch that began BEFORE an
    answer committed is still in flight, and publishing it raw republishes the
    answered question and resurrects its badge for a whole cycle.
    """
    _prune_answered()
    _snapshot["inbox"] = [q for q in rows if q.get("id") not in _answered_recently]
    _snapshot["inbox_count"] = len(_snapshot["inbox"])
    _snapshot["checked_at"] = now


def studio_snapshot():
    """A deep copy of the last refresher result. No I/O."""
    with _snapshot_lock:
        return copy.deepcopy(_snapshot)


def studio_poll_block(target, meta):
    """The `studio` key for /poll. Snapshot + cache reads only — no network.

    project comes from the active pane's cwd (same resolver as the ◇ deep
    link); effort comes only from a `studio_e<N>_s<M>` session name, because
    a cwd cannot tell us WHICH effort of a project a pane is working on.
    """
    snap = studio_snapshot()
    block = {
        "state": snap["state"],
        "inbox_count": snap["inbox_count"],
        "project_id": None,
        "project_name": None,
        "effort": None,
        "web_base": client.web_base(),
    }
    if snap["state"] != "connected":
        return block
    cwd = (meta or {}).get("project_dir")
    proj = _match_project(cwd, client.cached_projects()) if cwd else None
    if proj:
        block["project_id"] = proj["id"]
        block["project_name"] = proj.get("name")
    m = _SESSION_EFFORT_RE.match((target or "").split(":")[0])
    if m:
        block["effort"] = int(m.group(1))
    return block


def _explicitly_configured():
    """True when the user has actually pointed Assist at a Studio.

    The default loopback api_base is a convenience, not a configuration: a
    failed probe against it means "no Studio here" (unconfigured), not an
    outage (unreachable). A token, a web_base, or a non-default api_base are
    each an explicit act of configuration.
    """
    default_api = state.DEFAULT_SETTINGS["studio"]["api_base"].rstrip("/")
    try:
        web = str(state.get_setting("studio", "web_base") or "").strip()
    except (KeyError, TypeError):
        web = ""
    return bool(client.token()) or bool(web) or client.api_base() != default_api


def _state_for(code):
    if code == 200:
        return "connected"
    if code in (401, 403):
        return "unauthorized"
    if not client.configured() or not _explicitly_configured():
        return "unconfigured"
    return "unreachable"


def refresh_once():
    """One refresh cycle: probe, fetch the inbox, warm the projects cache.

    Returns the new connection state. Keeps the previous inbox on failure —
    stale data beats an empty panel during an outage.
    """
    with _refresh_lock:
        code = client.health()
        st = _state_for(code)
        inbox = None
        if st == "connected":
            icode, rows = client.inbox()
            if icode == 200:
                inbox = rows
            elif icode in (401, 403):
                st, code = "unauthorized", icode
            else:
                # ANY other inbox outcome (404/429/5xx/0) means the feature is
                # unusable even though /api/health answered. Falling through to
                # "connected" here would light a green state over a dead inbox.
                st, code = "unreachable", icode
            # Warm the projects cache so /poll's resolver never makes a call.
            client.projects()
        with _snapshot_lock:
            _snapshot["state"] = st
            _snapshot["http_code"] = code
            _snapshot["checked_at"] = time.time()
            if inbox is not None:
                _publish_inbox_locked(inbox, _snapshot["checked_at"])
        return st


def studio_refresher():
    """Background thread: the ONLY place Studio I/O happens on a timer."""
    next_health = 0.0
    while True:
        delay = _BACKOFF_INTERVAL
        try:
            # health is throttled to _HEALTH_INTERVAL while connected; a
            # disconnected Studio is re-probed every cycle so recovery is quick.
            now = time.time()
            with _snapshot_lock:
                connected = _snapshot["state"] == "connected"
            if connected and now < next_health:
                icode, rows = client.inbox()
                if icode == 200:
                    with _snapshot_lock:
                        # Same suppression as refresh_once(): this is the common
                        # path (every 10s while connected), so skipping the
                        # filter here would defeat it almost every time.
                        _publish_inbox_locked(rows, time.time())
                    delay = _INBOX_INTERVAL
                else:
                    delay = _INBOX_INTERVAL if refresh_once() == "connected" else _BACKOFF_INTERVAL
            else:
                next_health = now + _HEALTH_INTERVAL
                delay = _INBOX_INTERVAL if refresh_once() == "connected" else _BACKOFF_INTERVAL
        except Exception:
            delay = _BACKOFF_INTERVAL
        time.sleep(delay)


def _pane_cwd(target):
    """The active pane's cwd, or None if tmux can't be reached."""
    if not target:
        return None
    try:
        proc = subprocess.run(
            ["tmux", "display-message", "-t", target, "-p", "#{pane_current_path}"],
            capture_output=True, text=True, timeout=3,
        )
        if proc.returncode == 0 and proc.stdout.strip():
            return proc.stdout.strip()
    except Exception:
        pass
    return None


def _match_project(cwd, projects):
    """Best Studio project for a cwd: the one whose repo_path is the cwd or an
    ancestor of it. Longest repo_path wins (most specific). None if no match."""
    if not cwd:
        return None
    try:
        cwd_real = os.path.realpath(str(_find_project_dir(Path(cwd))))
    except Exception:
        cwd_real = os.path.realpath(cwd)
    best = None
    best_len = -1
    for p in projects:
        rp = p.get("repo_path")
        if not rp:
            continue
        rp_real = os.path.realpath(rp)
        if cwd_real == rp_real or cwd_real.startswith(rp_real + os.sep):
            if len(rp_real) > best_len:
                best, best_len = p, len(rp_real)
    return best


@studio_bp.route("/studio/link")
def studio_link():
    """Resolve a pane target to the Studio URL to open.

    matched project -> {web_base}/#/p/<id>;  otherwise -> {web_base}/#/.
    Always 200 with a usable url (home is the safe fallback)."""
    target = request.args.get("target", "")
    web = client.web_base()
    # cached_projects(): a tap must never block on Studio. The refresher warms
    # this cache every cycle while connected, so an empty cache means we are not
    # connected — and the Studio-home fallback below is the right answer then.
    proj = _match_project(_pane_cwd(target), client.cached_projects())
    if proj:
        return jsonify({
            "url": f"{web}/#/p/{proj['id']}",
            "project_id": proj["id"],
            "project_name": proj.get("name"),
        })
    return jsonify({"url": f"{web}/#/", "project_id": None, "project_name": None})


def _drop_from_snapshot(question_id):
    """Remove an answered question so the badge clears before the next refresh.

    Also remembers the id (see _answered_recently): an inbox fetch that started
    before this answer committed is still in flight and would otherwise
    republish the question, resurrecting it and its badge for a whole cycle.
    """
    with _snapshot_lock:
        _answered_recently[question_id] = time.time()
        _snapshot["inbox"] = [q for q in _snapshot["inbox"] if q.get("id") != question_id]
        _snapshot["inbox_count"] = len(_snapshot["inbox"])


@studio_bp.route("/studio/inbox")
def studio_inbox():
    """The attention queue, straight from the snapshot (no Studio I/O here)."""
    snap = studio_snapshot()
    return jsonify({
        "ok": True,
        "state": snap["state"],
        "checked_at": snap["checked_at"],
        "web_base": client.web_base(),
        "items": snap["inbox"],
    })


@studio_bp.route("/studio/answer", methods=["POST"])
def studio_answer():
    """Answer a blocking question, or accept a ✦ recommendation.

    Server-proxied by design: the API token stays in this process and the
    browser never learns it. Origin-gated like every other write (serve.py).

    action=answer -> submit {select, text} as the human's answer
    action=accept -> re-submit the stored recommendation as the answer
    """
    data = request.get_json(silent=True)
    if not isinstance(data, dict):
        return jsonify({"ok": False, "error": "object body required"}), 400
    qid = data.get("question_id")
    # NOT int(): bool is an int subclass and int(1.9) truncates to 1 — both would
    # cheerfully answer a different question than the caller named.
    if isinstance(qid, bool) or not isinstance(qid, int):
        return jsonify({"ok": False, "error": "question_id must be an integer"}), 400
    action = data.get("action", "answer")
    if action not in ("answer", "accept"):
        return jsonify({"ok": False, "error": "unknown action"}), 400

    item = next((q for q in studio_snapshot()["inbox"] if q.get("id") == qid), None)
    if item is None:
        return jsonify({"ok": False, "error": "question is not in the inbox"}), 404

    if action == "accept":
        # Accept is contracted for ✦ recommendations only. A pending blocking
        # question can carry a draft answer_json, and accepting that would
        # answer a question nobody proposed an answer for.
        if item.get("status") != "recommended":
            return jsonify({"ok": False, "error": "question is not a recommendation"}), 400
        answer_json = item.get("answer_json")
        if not isinstance(answer_json, dict) or not answer_json:
            return jsonify({"ok": False, "error": "no recommendation to accept"}), 400
    else:
        answer_json = {}
        select = data.get("select")
        if select is not None and not isinstance(select, str):
            return jsonify({"ok": False, "error": "select must be a string"}), 400
        if select:
            labels = item.get("options_json")
            labels = labels if isinstance(labels, list) else []
            # Mirrors the sto CLI: a label the question never offered is a loud
            # error, not a silently stored selection the SPA can't submit.
            if select not in labels:
                return jsonify({"ok": False, "error": "option not offered"}), 400
            answer_json["selections"] = [select]
        text = data.get("text")
        if text is not None and not isinstance(text, str):
            return jsonify({"ok": False, "error": "text must be a string"}), 400
        text = (text or "").strip()
        if text:
            answer_json["text"] = text
        if not answer_json:
            return jsonify({"ok": False, "error": "empty answer"}), 400

    code, record = client.answer(qid, answer_json, status="answered", actor="user")
    if code != 200:
        return jsonify({"ok": False, "error": f"Studio returned {code}"}), 502
    _drop_from_snapshot(qid)
    return jsonify({"ok": True, "question_id": qid, "record": record})


@studio_bp.route("/studio/test", methods=["POST"])
def studio_test():
    """Force a probe now — backs the Settings and connect-sheet Test buttons."""
    st = refresh_once()
    return jsonify({"ok": True, "state": st, "inbox_count": studio_snapshot()["inbox_count"]})

File: js/studio.js
// studio.js — Studio attention inbox panel, connect sheet, ◇ badge, effort chip.
// Every Studio call is server-proxied through /studio/* so the API token
// never reaches the browser (effort 268 decision).

var _studioPanelOpen = false;
var _studioSheetOpen = false;
var _studioInbox = [];
var _studioState = 'unconfigured';
var _studioWebBase = '';

var _STUDIO_STATE_LABEL = {
    connected: '● connected',
    unreachable: '○ unreachable',
    unauthorized: '⚠ unauthorized',
    unconfigured: '○ not connected',
};

// ================================================================
// Panel
// ================================================================
function toggleStudioPanel() {
    _studioPanelOpen = !_studioPanelOpen;
    document.getElementById('studio-panel').classList.toggle('visible', _studioPanelOpen);
    if (_studioPanelOpen) loadStudioInbox();
}

async function loadStudioInbox() {
    try {
        var r = await fetch('/studio/inbox');
        var d = await r.json();
        _studioInbox = d.items || [];
        _studioState = d.state || 'unconfigured';
        _studioWebBase = d.web_base || '';
    } catch (e) {
        // Keep the last good queue. Blanking it would render "Nothing needs you
        // right now" over a queue that still has items, because _studioState may
        // still say connected.
    }
    renderStudioInbox();
}

function renderStudioInbox() {
    var stateEl = document.getElementById('studio-state');
    if (stateEl) {
        stateEl.textContent = _STUDIO_STATE_LABEL[_studioState] || _studioState;
        stateEl.className = 'studio-state ' + _studioState;
    }
    var body = document.getElementById('studio-body');
    if (!body) return;
    body.innerHTML = '';

    if (_studioState !== 'connected') {
        var note = document.createElement('div');
        note.className = 'studio-empty';
        if (_studioState === 'unauthorized') {
            note.textContent = 'Studio rejected the API token — update it in Settings → Studio.';
        } else if (_studioState === 'unreachable') {
            note.textContent = 'Studio is not answering. Showing the last known queue.';
        } else {
            note.textContent = 'Not connected to Studio. Tap ◇ Studio to connect.';
        }
        body.appendChild(note);
        if (_studioState === 'unconfigured') return;
    }

    if (!_studioInbox.length) {
        var empty = document.createElement('div');
        empty.className = 'studio-empty';
        empty.textContent = 'Nothing needs you right now.';
        body.appendChild(empty);
        return;
    }
    for (var i = 0; i < _studioInbox.length; i++) {
        body.appendChild(_studioItemEl(_studioInbox[i]));
    }
}

// Built with createElement + textContent throughout: every field below is
// Studio-authored text, and this panel must not be an HTML injection path.
function _studioItemEl(q) {
    var isRec = q.status === 'recommended';
    var item = document.createElement('div');
    item.className = 'studio-item' + (isRec ? ' rec' : '');
    item.dataset.qid = q.id;

    var meta = document.createElement('div');
    meta.className = 'studio-meta';
    var proj = document.createElement('span');
    proj.className = 'studio-proj';
    proj.textContent = (q.project_slug || '?') + ' · e' + q.effort_id;
    meta.appendChild(proj);
    meta.appendChild(document.createTextNode(
        ' · Q' + q.id + (q.task_id ? ' · task ' + q.task_id : '')));
    item.appendChild(meta);

    var prompt = document.createElement('div');
    prompt.className = 'studio-prompt';
    if (isRec) {
        var pill = document.createElement('span');
        pill.className = 'studio-rec-pill';
        pill.textContent = '✦ recommended';
        prompt.appendChild(pill);
    }
    prompt.appendChild(document.createTextNode(q.prompt || ''));
    item.appendChild(prompt);

    if (q.evidence) {
        var ev = document.createElement('div');
        ev.className = 'studio-evidence';
        ev.textContent = '▸ ' + q.evidence;
        item.appendChild(ev);
    }

    if (isRec && q.answer_json) {
        var rec = document.createElement('div');
        rec.className = 'studio-rec-text';
        var parts = [];
        if (q.answer_json.selections) parts.push(q.answer_json.selections.join(', '));
        if (q.answer_json.text) parts.push(q.answer_json.text);
        if (q.answer_json.rationale) parts.push('— ' + q.answer_json.rationale);
        rec.textContent = parts.join('  ');
        item.appendChild(rec);
    }

    var opts = document.createElement('div');
    opts.className = 'studio-opts';

    if (isRec) {
        var accept = document.createElement('button');
        accept.className = 'studio-accept';
        accept.textContent = 'Accept ✦';
        accept.addEventListener('click', function () {
            studioAnswer(q.id, 'accept', null, null);
        });
        opts.appendChild(accept);
    }

    var labels = Array.isArray(q.options_json) ? q.options_json : [];
    labels.forEach(function (label) {
        var b = document.createElement('button');
        b.className = 'studio-opt';
        b.textContent = label;
        b.addEventListener('click', function () {
            studioAnswer(q.id, 'answer', label, null);
        });
        opts.appendChild(b);
    });

    var free = document.createElement('input');
    free.className = 'studio-freetext';
    free.placeholder = 'or type an answer…';
    free.addEventListener('keydown', function (e) {
        if (e.key === 'Enter' && free.value.trim()) {
            e.preventDefault();
            studioAnswer(q.id, 'answer', null, free.value.trim());
        }
    });
    opts.appendChild(free);

    var link = document.createElement('a');
    link.className = 'studio-open-link';
    link.textContent = 'open in Studio ↗';
    link.href = (_studioWebBase || '') + '/#/p/' + q.project_id;
    link.target = '_blank';
    link.rel = 'noopener';
    opts.appendChild(link);

    item.appendChild(opts);
    return item;
}

async function studioAnswer(qid, action, select, text) {
    // A phone double-tap would otherwise fire twice: the second POST 404s (the
    // question is already out of the snapshot) and its error flash overwrites
    // the success one.
    var itemEl = document.querySelector('.studio-item[data-qid="' + qid + '"]');
    var controls = itemEl ? itemEl.querySelectorAll('button, input') : [];
    var reenable = function () {
        Array.prototype.forEach.call(controls, function (c) { c.disabled = false; });
    };
    Array.prototype.forEach.call(controls, function (c) { c.disabled = true; });
    var payload = { question_id: qid, action: action };
    if (select) payload.select = select;
    if (text) payload.text = text;
    try {
        var r = await fetch('/studio/answer', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload),
        });
        var d = await r.json();
        if (d.ok) {
            showFlash('ok', action === 'accept' ? 'Recommendation accepted' : 'Answered Q' + qid);
            loadStudioInbox();
        } else {
            reenable();
            showFlash('error', d.error || 'Studio answer failed');
        }
    } catch (e) {
        reenable();
        showFlash('error', 'Offline');
    }
}

// ================================================================
// Poll integration — badge, button state, effort chip
// ================================================================
function _applyStudioData(block) {
    if (!block) return;
    _studioState = block.state || 'unconfigured';
    if (block.web_base) _studioWebBase = block.web_base;

    var btn = document.getElementById('btn-studio');
    var count = block.inbox_count || 0;
    if (btn) {
        btn.classList.toggle('grey', _studioState === 'unreachable' || _studioState === 'unconfigured');
        btn.classList.toggle('locked', _studioState === 'unauthorized');
        // Lock glyph and count badge are mutually exclusive: a rejected token
        // means the count on screen is stale, so don't show a number at all.
        var lock = document.getElementById('studio-lock');
        if (lock) lock.classList.toggle('hidden', _studioState !== 'unauthorized');
        var badge = document.getElementById('studio-badge');
        if (badge) {
            badge.textContent = count > 9 ? '9+' : String(count);
            badge.classList.toggle('hidden', !(count > 0 && _studioState === 'connected'));
        }
    }
    // Re-read the snapshot on EVERY poll while the panel is open — not only
    // when the count changes. An outage preserves inbox_count (the server keeps
    // serving the last snapshot), so a count-gated refresh would leave the
    // header reading "connected" for the entire outage. This is an Assist-local
    // read; the refresher thread still owns every call to Studio itself.
    if (_studioPanelOpen) loadStudioInbox();
    updateStudioChip(block);
}

function updateStudioChip(block) {
    var chip = document.getElementById('ci-effort');
    if (!chip) return;
    if (_studioState !== 'connected' || (!block.project_id && !block.effort)) {
        chip.classList.add('hidden');
        chip.textContent = '';
        return;
    }
    var label = '◇ ' + (block.project_name || 'Studio');
    if (block.effort) label += ' · e' + block.effort;
    chip.textContent = label;
    chip.classList.remove('hidden');
    chip.onclick = function () {
        if (!block.project_id) return;
        window.open((_studioWebBase || '') + '/#/p/' + block.project_id, '_blank', 'noopener');
    };
}

// ================================================================
// Connect sheet — the free -> connected doorway
// ================================================================
function openStudioConnect() {
    _studioSheetOpen = true;
    var sheet = document.getElementById('studio-sheet');
    sheet.classList.add('visible');
    // The markup ships aria-hidden="true"; leaving it set would keep the open
    // sheet invisible to assistive tech.
    sheet.setAttribute('aria-hidden', 'false');
    var s = (window.SETTINGS && SETTINGS.studio) || {};
    document.getElementById('studio-web-base').value = s.web_base || '';
    document.getElementById('studio-api-base').value = s.api_base || '';
    document.getElementById('studio-api-token').value = '';   // never seed the mask
    document.getElementById('studio-connect-result').textContent = '';
}

function closeStudioConnect() {
    _studioSheetOpen = false;
    var sheet = document.getElementById('studio-sheet');
    sheet.classList.remove('visible');
    sheet.setAttribute('aria-hidden', 'true');
}

async function studioConnectSave() {
    var out = document.getElementById('studio-connect-result');
    var patch = { studio: {
        web_base: document.getElementById('studio-web-base').value.trim(),
        api_base: document.getElementById('studio-api-base').value.trim(),
    } };
    var tok = document.getElementById('studio-api-token').value.trim();
    if (tok) patch.studio.api_token = tok;
    out.textContent = 'Testing…';
    try {
        var r = await fetch('/api/settings', {
            method: 'PATCH',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(patch),
        });
        var d = await r.json();
        if (!d.ok) { out.textContent = d.error || 'Save failed'; return; }
        SETTINGS = d.settings;
        var t = await fetch('/studio/test', { method: 'POST' });
        var td = await t.json();
        _studioState = td.state;
        out.textContent = _STUDIO_STATE_LABEL[td.state] || td.state;
        if (td.state === 'connected') {
            showFlash('ok', 'Studio connected');
            closeStudioConnect();
            loadStudioInbox();
        }
    } catch (e) {
        out.textContent = 'Offline';
    }
}

async function studioTestConnection() {
    try {
        var r = await fetch('/studio/test', { method: 'POST' });
        var d = await r.json();
        _studioState = d.state;
        showFlash(d.state === 'connected' ? 'ok' : 'error',
            'Studio: ' + (_STUDIO_STATE_LABEL[d.state] || d.state));
        if (_studioPanelOpen) loadStudioInbox();
    } catch (e) {
        showFlash('error', 'Offline');
    }
}
